Crashing with CUDA 8.0 and Tensorflow 1.4

Just reporting a solved issue: after 6d59402fbe60814792e19c7a9216f2097edd90d7, RETURNN crashes with CUDA 8.0 (Tensorflow 1.4, and Python 3.6).

Using a newer version of CUDA solves the issue (I'm using CUDA 9.0 and Tensoflow 1.8).

Here is the error message for the record:
```
OpCodeCompiler call: /usr/local/cuda-8.0/bin/nvcc -shared -O2 -std=c++11 -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public -I /usr/local/cuda-8.0/include -L /usr/local/cuda-8.0/lib64 -x cu -v -DGOOGLE_CUDA=1 -Xcompiler -fPIC -Xcompiler -v -arch compute_61 -D_GLIBCXX_USE_CXX11_ABI=0 -g /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc -o /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.so -L/u/irie/.local/lib/python3.6/site-packages/numpy/.libs -l:libopenblasp-r0-8dca6697.3.0.dev.so -L/u/irie/tools/anaconda3/lib -l:libmkl_rt.so -L/u/irie/.local/lib/python3.6/site-packages/tensorflow -ltensorflow_framework
OpCodeCompiler: /usr/local/cuda-8.0/bin/nvcc failed.
Original stdout/stderr:
Using built-in specs.
COLLECT_GCC=gcc
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.10' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10) 
COLLECT_GCC_OPTIONS='-std=c++11' '-D' '__CUDA_ARCH__=610' '-E' '-D' 'CUDA_DOUBLE_MATH_FUNCTIONS' '-D' '__CUDACC__' '-D' '__NVCC__' '-fPIC' '-v' '-O2' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public' '-I' '/usr/local/cuda-8.0/include' '-I' '/usr/local/cuda-8.0/bin/..//include' '-D' '__CUDACC_VER__=80044' '-D' '__CUDACC_VER_BUILD__=44' '-D' '__CUDACC_VER_MINOR__=0' '-D' '__CUDACC_VER_MAJOR__=8' '-D' 'GOOGLE_CUDA=1' '-D' '_GLIBCXX_USE_CXX11_ABI=0' '-include' 'cuda_runtime.h' '-m64' '-g' '-gdwarf-2' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/5/cc1plus -E -quiet -v -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public -I /usr/local/cuda-8.0/include -I /usr/local/cuda-8.0/bin/..//include -imultiarch x86_64-linux-gnu -D_GNU_SOURCE -D __CUDA_ARCH__=610 -D CUDA_DOUBLE_MATH_FUNCTIONS -D __CUDACC__ -D __NVCC__ -D __CUDACC_VER__=80044 -D __CUDACC_VER_BUILD__=44 -D __CUDACC_VER_MINOR__=0 -D __CUDACC_VER_MAJOR__=8 -D GOOGLE_CUDA=1 -D _GLIBCXX_USE_CXX11_ABI=0 -include cuda_runtime.h /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc -m64 -mtune=generic -march=x86-64 -std=c++11 -fPIC -g -gdwarf-2 -fworking-directory -O2 -fstack-protector-strong -Wformat -Wformat-security
ignoring duplicate directory "/usr/include/x86_64-linux-gnu/c++/5"
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/5/../../../../x86_64-linux-gnu/include"
ignoring duplicate directory "/usr/local/cuda-8.0/bin/..//include"
#include "..." search starts here:
#include <...> search starts here:
 /u/irie/.local/lib/python3.6/site-packages/tensorflow/include
 /u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public
 /usr/local/cuda-8.0/include
 /usr/include/c++/5
 /usr/include/x86_64-linux-gnu/c++/5
 /usr/include/c++/5/backward
 /usr/lib/gcc/x86_64-linux-gnu/5/include
 /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
In file included from /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc:285:0:
/u/irie/src/head_rt/NativeOp.cpp:318:0: warning: "elem_atomic_add" redefined
 #define elem_atomic_add(x, v) atomicAdd(x, v)
 ^
In file included from /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc:36:0:
/u/irie/src/head_rt/NativeOp.cpp:431:0: note: this is the location of the previous definition
 #define elem_atomic_add(x, v) (*x += v)  // ignore atomic for now...
 ^
COMPILER_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/
LIBRARY_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../../lib/:/lib/x86_64-linux-gnu/:/lib/../lib/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-std=c++11' '-D' '__CUDA_ARCH__=610' '-E' '-D' 'CUDA_DOUBLE_MATH_FUNCTIONS' '-D' '__CUDACC__' '-D' '__NVCC__' '-fPIC' '-v' '-O2' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public' '-I' '/usr/local/cuda-8.0/include' '-I' '/usr/local/cuda-8.0/bin/..//include' '-D' '__CUDACC_VER__=80044' '-D' '__CUDACC_VER_BUILD__=44' '-D' '__CUDACC_VER_MINOR__=0' '-D' '__CUDACC_VER_MAJOR__=8' '-D' 'GOOGLE_CUDA=1' '-D' '_GLIBCXX_USE_CXX11_ABI=0' '-include' 'cuda_runtime.h' '-m64' '-g' '-gdwarf-2' '-mtune=generic' '-march=x86-64'
#$ cudafe --allow_managed --m64 --gnu_version=50400 --c++11 -tused --no_remove_unneeded_entities  --gen_c_file_name "/var/tmp/tmpxft_0000223f_00000000-4_LstmGenericBase.cudafe1.c" --stub_file_name "/var/tmp/tmpxft_0000223f_00000000-4_LstmGenericBase.cudafe1.stub.c" --gen_device_file_name "/var/tmp//tmpxft_0000223f_00000000-4_LstmGenericBase.cudafe1.gpu" --nv_arch "compute_61" --gen_module_id_file --module_id_file_name "/var/tmp//tmpxft_0000223f_00000000-3_LstmGenericBase.module_id" --include_file_name "tmpxft_0000223f_00000000-2_LstmGenericBase.fatbin.c" "/var/tmp//tmpxft_0000223f_00000000-9_LstmGenericBase.cpp1.ii" 
/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/google/protobuf/arena_impl.h(52): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/google/protobuf/arena_impl.h(147): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1265): warning: calling a constexpr __host__ function("real") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1265): warning: calling a constexpr __host__ function("imag") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1265): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1265): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1270): warning: calling a constexpr __host__ function("real") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1270): warning: calling a constexpr __host__ function("imag") from a __host__ __device__ function("abs") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1270): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/../../../Eigen/src/Core/MathFunctions.h(1270): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h(133): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h(138): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h(208): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/unsupported/Eigen/CXX11/src/Tensor/TensorRandom.h(213): warning: calling a constexpr __host__ function from a __host__ __device__ function is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/tensorflow/core/framework/function.h(476): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/tensorflow/core/framework/function.h(477): warning: integer conversion resulted in a change of sign

/u/irie/src/head_rt/NativeOp.cpp(723): warning: variable "shiftA" was declared but never referenced

/u/irie/src/head_rt/NativeOp.cpp(723): warning: variable "shiftA" was declared but never referenced

#$ gcc -std=c++11 -E -x c++ -D__CUDACC__ -D__NVCC__  -fPIC -v -O2 -I"/u/irie/.local/lib/python3.6/site-packages/tensorflow/include" -I"/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public" -I"/usr/local/cuda-8.0/include" "-I/usr/local/cuda-8.0/bin/..//include"   -D"__CUDACC_VER__=80044" -D"__CUDACC_VER_BUILD__=44" -D"__CUDACC_VER_MINOR__=0" -D"__CUDACC_VER_MAJOR__=8" -D"GOOGLE_CUDA=1" -D"_GLIBCXX_USE_CXX11_ABI=0" -include "cuda_runtime.h" -m64 -g -gdwarf-2 "/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc" > "/var/tmp//tmpxft_0000223f_00000000-5_LstmGenericBase.cpp4.ii" 
Using built-in specs.
COLLECT_GCC=gcc
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.10' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10) 
COLLECT_GCC_OPTIONS='-std=c++11' '-E' '-D' '__CUDACC__' '-D' '__NVCC__' '-fPIC' '-v' '-O2' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public' '-I' '/usr/local/cuda-8.0/include' '-I' '/usr/local/cuda-8.0/bin/..//include' '-D' '__CUDACC_VER__=80044' '-D' '__CUDACC_VER_BUILD__=44' '-D' '__CUDACC_VER_MINOR__=0' '-D' '__CUDACC_VER_MAJOR__=8' '-D' 'GOOGLE_CUDA=1' '-D' '_GLIBCXX_USE_CXX11_ABI=0' '-include' 'cuda_runtime.h' '-m64' '-g' '-gdwarf-2' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/5/cc1plus -E -quiet -v -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include -I /u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public -I /usr/local/cuda-8.0/include -I /usr/local/cuda-8.0/bin/..//include -imultiarch x86_64-linux-gnu -D_GNU_SOURCE -D __CUDACC__ -D __NVCC__ -D __CUDACC_VER__=80044 -D __CUDACC_VER_BUILD__=44 -D __CUDACC_VER_MINOR__=0 -D __CUDACC_VER_MAJOR__=8 -D GOOGLE_CUDA=1 -D _GLIBCXX_USE_CXX11_ABI=0 -include cuda_runtime.h /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc -m64 -mtune=generic -march=x86-64 -std=c++11 -fPIC -g -gdwarf-2 -fworking-directory -O2 -fstack-protector-strong -Wformat -Wformat-security
ignoring duplicate directory "/usr/include/x86_64-linux-gnu/c++/5"
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/5/../../../../x86_64-linux-gnu/include"
ignoring duplicate directory "/usr/local/cuda-8.0/bin/..//include"
#include "..." search starts here:
#include <...> search starts here:
 /u/irie/.local/lib/python3.6/site-packages/tensorflow/include
 /u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public
 /usr/local/cuda-8.0/include
 /usr/include/c++/5
 /usr/include/x86_64-linux-gnu/c++/5
 /usr/include/c++/5/backward
 /usr/lib/gcc/x86_64-linux-gnu/5/include
 /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
In file included from /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc:285:0:
/u/irie/src/head_rt/NativeOp.cpp:318:0: warning: "elem_atomic_add" redefined
 #define elem_atomic_add(x, v) atomicAdd(x, v)
 ^
In file included from /var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc:36:0:
/u/irie/src/head_rt/NativeOp.cpp:431:0: note: this is the location of the previous definition
 #define elem_atomic_add(x, v) (*x += v)  // ignore atomic for now...
 ^
COMPILER_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/
LIBRARY_PATH=/usr/lib/gcc/x86_64-linux-gnu/5/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../../lib/:/lib/x86_64-linux-gnu/:/lib/../lib/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/5/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-std=c++11' '-E' '-D' '__CUDACC__' '-D' '__NVCC__' '-fPIC' '-v' '-O2' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include' '-I' '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public' '-I' '/usr/local/cuda-8.0/include' '-I' '/usr/local/cuda-8.0/bin/..//include' '-D' '__CUDACC_VER__=80044' '-D' '__CUDACC_VER_BUILD__=44' '-D' '__CUDACC_VER_MINOR__=0' '-D' '__CUDACC_VER_MAJOR__=8' '-D' 'GOOGLE_CUDA=1' '-D' '_GLIBCXX_USE_CXX11_ABI=0' '-include' 'cuda_runtime.h' '-m64' '-g' '-gdwarf-2' '-mtune=generic' '-march=x86-64'
#$ cudafe++ --allow_managed --m64 --gnu_version=50400 --c++11 --parse_templates  --gen_c_file_name "/var/tmp/tmpxft_0000223f_00000000-4_LstmGenericBase.cudafe1.cpp" --stub_file_name "tmpxft_0000223f_00000000-4_LstmGenericBase.cudafe1.stub.c" --module_id_file_name "/var/tmp//tmpxft_0000223f_00000000-3_LstmGenericBase.module_id" "/var/tmp//tmpxft_0000223f_00000000-5_LstmGenericBase.cpp4.ii" 
/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/google/protobuf/arena_impl.h(52): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/google/protobuf/arena_impl.h(147): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/google/protobuf/generated_message_reflection.h(689): warning: variable "unused" was set but never used

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/tensorflow/core/framework/function.h(476): warning: integer conversion resulted in a change of sign

/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/tensorflow/core/framework/function.h(477): warning: integer conversion resulted in a change of sign

/u/irie/src/head_rt/NativeOp.cpp(211): error: qualified name is not allowed

/u/irie/src/head_rt/NativeOp.cpp(211): error: expected a ";"

/u/irie/src/head_rt/NativeOp.cpp(212): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(213): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(214): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(215): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(216): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(217): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(243): error: pointer to incomplete class type is not allowed

/u/irie/src/head_rt/NativeOp.cpp(244): error: identifier "get_transpose" is undefined

/u/irie/src/head_rt/NativeOp.cpp(256): error: pointer to incomplete class type is not allowed

/u/irie/src/head_rt/NativeOp.cpp(723): warning: variable "shiftA" was declared but never referenced

/u/irie/src/head_rt/NativeOp.cpp(211): error: qualified name is not allowed

/u/irie/src/head_rt/NativeOp.cpp(211): error: expected a ";"

/u/irie/src/head_rt/NativeOp.cpp(212): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(213): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(214): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(215): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(216): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(217): error: identifier "DeviceMemoryType" is undefined

/u/irie/src/head_rt/NativeOp.cpp(243): error: pointer to incomplete class type is not allowed

/u/irie/src/head_rt/NativeOp.cpp(244): error: identifier "get_transpose" is undefined

/u/irie/src/head_rt/NativeOp.cpp(256): error: pointer to incomplete class type is not allowed

/u/irie/src/head_rt/NativeOp.cpp(723): warning: variable "shiftA" was declared but never referenced

22 errors detected in the compilation of "/var/tmp//tmpxft_0000223f_00000000-5_LstmGenericBase.cpp4.ii".
# --error 0x2 --


Exception creating layer root/'lstm0' of class RecLayer with opts:
{'L2': 0.0,
 'bias_init': 'random_normal_initializer(mean=0.0, stddev=0.1)',
 'direction': 1,
 'dropout': 0.5,
 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)',
 'n_out': 1024,
 'name': 'lstm0',
 'network': <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>,
 'output': Data(name='lstm0_output', shape=(None, 1024), batch_dim_axis=1),
 'recurrent_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)',
 'sources': [<LinearLayer 'input' out_type=Data(shape=(None, 128))>],
 'unit': 'lstm'}

EXCEPTION
Traceback (most recent call last):
  File "returnn/rnn.py", line 595, in <module>
    line: main(sys.argv)
    locals:
      main = <local> <function main at 0x7fc46c9e49d8>
      sys = <local> <module 'sys' (built-in)>
      sys.argv = <local> ['returnn/rnn.py', 'config-train/example3.config'], _[0]: {len = 14}
  File "returnn/rnn.py", line 583, in main
    line: executeMainTask()
    locals:
      executeMainTask = <global> <function executeMainTask at 0x7fc46c9e48c8>
  File "returnn/rnn.py", line 414, in executeMainTask
    line: engine.init_train_from_config(config, train_data, dev_data, eval_data)
    locals:
      engine = <global> <TFEngine.Engine object at 0x7fc414c71320>
      engine.init_train_from_config = <global> <bound method Engine.init_train_from_config of <TFEngine.Engine object at 0x7fc414c71320>>
      config = <global> <Config.Config object at 0x7fc478a345c0>
      train_data = <global> <LmDataset 'train' epoch=1>
      dev_data = <global> <LmDataset 'dev' epoch=1>
      eval_data = <global> None
  File "/u/irie/src/head_rt/TFEngine.py", line 895, in init_train_from_config
    line: self.init_network_from_config(config)
    locals:
      self = <local> <TFEngine.Engine object at 0x7fc414c71320>
      self.init_network_from_config = <local> <bound method Engine.init_network_from_config of <TFEngine.Engine object at 0x7fc414c71320>>
      config = <local> <Config.Config object at 0x7fc478a345c0>
  File "/u/irie/src/head_rt/TFEngine.py", line 936, in init_network_from_config
    line: self._init_network(net_desc=net_dict, epoch=self.epoch)
    locals:
      self = <local> <TFEngine.Engine object at 0x7fc414c71320>
      self._init_network = <local> <bound method Engine._init_network of <TFEngine.Engine object at 0x7fc414c71320>>
      net_desc = <not found>
      net_dict = <local> {'input': {'class': 'linear', 'n_out': 128, 'activation': 'identity', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'from': ['data:delayed']}, 'lstm0': {'class': 'rec', 'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurre...
      epoch = <local> None
      self.epoch = <local> 1
  File "/u/irie/src/head_rt/TFEngine.py", line 1060, in _init_network
    line: self.network, self.updater = self.create_network(
            config=self.config,
            rnd_seed=net_random_seed,
            train_flag=train_flag, eval_flag=self.use_eval_flag, search_flag=self.use_search_flag,
            initial_learning_rate=getattr(self, "initial_learning_rate", None),
            net_dict=net_desc)
    locals:
      self = <local> <TFEngine.Engine object at 0x7fc414c71320>
      self.network = <local> None
      self.updater = <local> None
      self.create_network = <local> <bound method Engine.create_network of <class 'TFEngine.Engine'>>
      config = <not found>
      self.config = <local> <Config.Config object at 0x7fc478a345c0>
      rnd_seed = <not found>
      net_random_seed = <local> 1
      train_flag = <local> <tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>
      eval_flag = <not found>
      self.use_eval_flag = <local> True
      search_flag = <not found>
      self.use_search_flag = <local> False
      initial_learning_rate = <not found>
      getattr = <builtin> <built-in function getattr>
      net_dict = <not found>
      net_desc = <local> {'input': {'class': 'linear', 'n_out': 128, 'activation': 'identity', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'from': ['data:delayed']}, 'lstm0': {'class': 'rec', 'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurre...
  File "/u/irie/src/head_rt/TFEngine.py", line 1091, in create_network
    line: network.construct_from_dict(net_dict)
    locals:
      network = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      network.construct_from_dict = <local> <bound method TFNetwork.construct_from_dict of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'input': {'class': 'linear', 'n_out': 128, 'activation': 'identity', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'from': ['data:delayed']}, 'lstm0': {'class': 'rec', 'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurre...
  File "/u/irie/src/head_rt/TFNetwork.py", line 338, in construct_from_dict
    line: self.construct_layer(net_dict, name)
    locals:
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'input': {'class': 'linear', 'n_out': 128, 'activation': 'identity', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'from': ['data:delayed']}, 'lstm0': {'class': 'rec', 'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurre...
      name = <local> 'output', len = 6
  File "/u/irie/src/head_rt/TFNetwork.py", line 415, in construct_layer
    line: layer_class.transform_config_dict(layer_desc, network=self, get_layer=get_layer)
    locals:
      layer_class = <local> <class 'TFNetworkLayer.SoftmaxLayer'>
      layer_class.transform_config_dict = <local> <bound method LayerBase.transform_config_dict of <class 'TFNetworkLayer.SoftmaxLayer'>>
      layer_desc = <local> {'dropout': 0.5, 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'bias_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'loss': 'ce', 'target': 'data'}
      network = <not found>
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      get_layer = <local> <function TFNetwork.construct_layer.<locals>.get_layer at 0x7fc3f074eb70>
  File "/u/irie/src/head_rt/TFNetworkLayer.py", line 358, in transform_config_dict
    line: for src_name in src_names
    locals:
      src_name = <not found>
      src_names = <local> ['lstm0']
  File "/u/irie/src/head_rt/TFNetworkLayer.py", line 359, in <listcomp>
    line: d["sources"] = [
            get_layer(src_name)
            for src_name in src_names
            if not src_name == "none"]
    locals:
      d = <not found>
      get_layer = <local> <function TFNetwork.construct_layer.<locals>.get_layer at 0x7fc3f074eb70>
      src_name = <local> 'lstm0'
      src_names = <not found>
  File "/u/irie/src/head_rt/TFNetwork.py", line 388, in get_layer
    line: return self.construct_layer(net_dict=net_dict, name=src_name)  # set get_layer to wrap construct_layer
    locals:
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self.construct_layer = <local> <bound method TFNetwork.construct_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      net_dict = <local> {'input': {'class': 'linear', 'n_out': 128, 'activation': 'identity', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'from': ['data:delayed']}, 'lstm0': {'class': 'rec', 'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurre...
      name = <not found>
      src_name = <local> 'lstm0'
  File "/u/irie/src/head_rt/TFNetwork.py", line 418, in construct_layer
    line: return add_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      add_layer = <local> <bound method TFNetwork.add_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'lstm0'
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurrent_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'bias_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'n_out': 1024, 'dropout': 0.5, 'L2': 0.0, 'direction': 1, 'sourc..., len = 9
  File "/u/irie/src/head_rt/TFNetwork.py", line 507, in add_layer
    line: layer = self._create_layer(name=name, layer_class=layer_class, **layer_desc)
    locals:
      layer = <not found>
      self = <local> <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>
      self._create_layer = <local> <bound method TFNetwork._create_layer of <TFNetwork 'root' train=<tf.Tensor 'globals/train_flag:0' shape=() dtype=bool>>>
      name = <local> 'lstm0'
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurrent_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'bias_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'n_out': 1024, 'dropout': 0.5, 'L2': 0.0, 'direction': 1, 'sourc..., len = 9
  File "/u/irie/src/head_rt/TFNetwork.py", line 464, in _create_layer
    line: layer = layer_class(**layer_desc)
    locals:
      layer = <not found>
      layer_class = <local> <class 'TFNetworkRecLayer.RecLayer'>
      layer_desc = <local> {'unit': 'lstm', 'forward_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'recurrent_weights_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'bias_init': 'random_normal_initializer(mean=0.0, stddev=0.1)', 'n_out': 1024, 'dropout': 0.5, 'L2': 0.0, 'direction': 1, 'sourc..., len = 12
  File "/u/irie/src/head_rt/TFNetworkRecLayer.py", line 170, in __init__
    line: self.cell = self._get_cell(unit, unit_opts=unit_opts)
    locals:
      self = <local> <RecLayer 'lstm0' out_type=Data(shape=(None, 1024), batch_dim_axis=1)>
      self.cell = <local> !AttributeError: 'RecLayer' object has no attribute 'cell'
      self._get_cell = <local> <bound method RecLayer._get_cell of <RecLayer 'lstm0' out_type=Data(shape=(None, 1024), batch_dim_axis=1)>>
      unit = <local> 'lstm'
      unit_opts = <local> None
  File "/u/irie/src/head_rt/TFNetworkRecLayer.py", line 468, in _get_cell
    line: cell = rnn_cell_class(
            n_hidden=n_hidden, n_input_dim=self.input_data.dim,
            input_is_sparse=self.input_data.sparse,
            step=self._direction, **unit_opts)
    locals:
      cell = <not found>
      rnn_cell_class = <local> <class 'TFNativeOp.NativeLstmCell'>
      n_hidden = <local> 1024
      n_input_dim = <not found>
      self = <local> <RecLayer 'lstm0' out_type=Data(shape=(None, 1024), batch_dim_axis=1)>
      self.input_data = <local> Data(name='input_output', shape=(None, 128))
      self.input_data.dim = <local> 128
      input_is_sparse = <not found>
      self.input_data.sparse = <local> False
      step = <not found>
      self._direction = <local> 1
      unit_opts = <local> {}
  File "/u/irie/src/head_rt/TFNativeOp.py", line 603, in __init__
    line: self.op = make_lstm_op()
    locals:
      self = <local> <TFNativeOp.NativeLstmCell object at 0x7fc3e11d83c8>
      self.op = <local> !AttributeError: 'NativeLstmCell' object has no attribute 'op'
      make_lstm_op = <global> <function make_lstm_op at 0x7fc3e11d71e0>
  File "/u/irie/src/head_rt/TFNativeOp.py", line 556, in make_lstm_op
    line: return make_op(NativeOp.LstmGenericBase, **kwargs)
    locals:
      make_op = <global> <function make_op at 0x7fc3e11d7158>
      NativeOp = <global> <module 'NativeOp' from '/u/irie/src/head_rt/NativeOp.py'>
      NativeOp.LstmGenericBase = <global> <class 'NativeOp.LstmGenericBase'>
      kwargs = <local> {}
  File "/u/irie/src/head_rt/TFNativeOp.py", line 546, in make_op
    line: return maker.make_op()
    locals:
      maker = <local> <TFNativeOp.OpMaker object at 0x7fc3e11d85f8>
      maker.make_op = <local> <bound method OpMaker.make_op of <TFNativeOp.OpMaker object at 0x7fc3e11d85f8>>
  File "/u/irie/src/head_rt/TFNativeOp.py", line 461, in make_op
    line: mod = self._make_mod()
    locals:
      mod = <not found>
      self = <local> <TFNativeOp.OpMaker object at 0x7fc3e11d85f8>
      self._make_mod = <local> <bound method OpMaker._make_mod of <TFNativeOp.OpMaker object at 0x7fc3e11d85f8>>
  File "/u/irie/src/head_rt/TFNativeOp.py", line 452, in _make_mod
    line: mod = comp.load_tf_module()
    locals:
      mod = <not found>
      comp = <local> <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>
      comp.load_tf_module = <local> <bound method OpCodeCompiler.load_tf_module of <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>>
  File "/u/irie/src/head_rt/TFUtil.py", line 3326, in load_tf_module
    line: self._maybe_compile()
    locals:
      self = <local> <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>
      self._maybe_compile = <local> <bound method NativeCodeCompiler._maybe_compile of <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>>
  File "/u/irie/src/head_rt/Util.py", line 2400, in _maybe_compile
    line: self._maybe_compile_inner()
    locals:
      self = <local> <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>
      self._maybe_compile_inner = <local> <bound method NativeCodeCompiler._maybe_compile_inner of <OpCodeCompiler 'LstmGenericBase' in '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423'>>
  File "/u/irie/src/head_rt/Util.py", line 2448, in _maybe_compile_inner
    line: raise CalledProcessError(returncode=proc.returncode, cmd=cmd_args)
    locals:
      CalledProcessError = <local> <class 'subprocess.CalledProcessError'>
      returncode = <not found>
      proc = <local> <subprocess.Popen object at 0x7fc3e11d8748>
      proc.returncode = <local> 2
      cmd = <global> <function cmd at 0x7fc455ca27b8>
      cmd_args = <local> ['/usr/local/cuda-8.0/bin/nvcc', '-shared', '-O2', '-std=c++11', '-I', '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include', '-I', '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public', '-I', '/usr/local/cuda-8.0/include', '-L', '/usr/local/cuda-8.0/lib6..., len = 33, _[0]: {len = 28}
CalledProcessError: Command '['/usr/local/cuda-8.0/bin/nvcc', '-shared', '-O2', '-std=c++11', '-I', '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include', '-I', '/u/irie/.local/lib/python3.6/site-packages/tensorflow/include/external/nsync/public', '-I', '/usr/local/cuda-8.0/include', '-L', '/usr/local/cuda-8.0/lib64', '-x', 'cu', '-v', '-DGOOGLE_CUDA=1', '-Xcompiler', '-fPIC', '-Xcompiler', '-v', '-arch', 'compute_61', '-D_GLIBCXX_USE_CXX11_ABI=0', '-g', '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.cc', '-o', '/var/tmp//irie/returnn_tf_cache/ops/LstmGenericBase/061d404423/LstmGenericBase.so', '-L/u/irie/.local/lib/python3.6/site-packages/numpy/.libs', '-l:libopenblasp-r0-8dca6697.3.0.dev.so', '-L/u/irie/tools/anaconda3/lib', '-l:libmkl_rt.so', '-L/u/irie/.local/lib/python3.6/site-packages/tensorflow', '-ltensorflow_framework']' returned non-zero exit status 2.
Unhandled exception <class 'subprocess.CalledProcessError'> in thread <_MainThread(MainThread, started 140481815631616)>, proc 8015.
```

--- a/NativeOp.cpp
+++ b/NativeOp.cpp
@@ -95,8 +95,10 @@ Ndarray* Ndarray_Copy(const Ndarray* self) {
 
 #if (TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION >= 5) || (TF_MAJOR_VERSION > 1)
 #define TF_issue_6602_workaround 0
+#define TWOD_LSTM_SUPPORT 1
 #else
 #define TF_issue_6602_workaround 1
+#define TWOD_LSTM_SUPPORT 0
 #endif
 
 #if TF_issue_6602_workaround
@@ -208,6 +210,7 @@ static void tf_cuda_sgemm_batched(
     T beta = *beta_;
 // https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/rnn/kernels/blas_gemm.cc
 #if GOOGLE_CUDA
+#if TWOD_LSTM_SUPPORT
     typedef perftools::gputools::DeviceMemory<float> DeviceMemoryType;
     std::vector<DeviceMemoryType> a_device_memory;
     std::vector<DeviceMemoryType> b_device_memory;
@@ -258,6 +261,9 @@ static void tf_cuda_sgemm_batched(
                 .ok();
         OP_REQUIRES(context, blas_launch_status, errors::Aborted("BlockHostUntilDone failed!"));
     }
+#else  // TWOD_LSTM_SUPPORT
+    context->SetStatus(errors::InvalidArgument("For 2D-LSTMs, TensorFlow 1.5 or later is required"));
+#endif
 #else  // GOOGLE_CUDA
     context->SetStatus(errors::InvalidArgument("CuBlasGemm needs CUDA."));
 #endif  // GOOGLE_CUDA

